File: python/list-ops/list_ops.py
def append(list1, list2):
    return list1 + list2


def concat(lists):
    result = []
    for item in lists:
        result += item
    return result


def filter(function, input_list):
    return [item for item in input_list if function(item)]


def length(input_list):
    return sum(1 for item in input_list)


def map(function, input_list):
    return [function(item) for item in input_list ]


def foldl(function, input_list, initial):
    result = initial
    for item in input_list:
        result = function(result, item)
    return result


def foldr(function, input_list, initial):
    return foldl(function, input_list[::-1], initial)


def reverse(input_list):
    return input_list[::-1]

File: fsharp/list-ops/ListOps.fs
module ListOps

let rec foldl folder state list =
    match list with
    | [] -> state
    | h :: t -> foldl folder (folder state h) t

let rec foldr folder state list = 
    match list with
    | [] -> state
    | h :: t -> folder (foldr folder state t) h

let length list =
    foldl (fun acc el -> 1 + acc) 0 list

let reverse list =
    foldl (fun acc el -> el :: acc) [] list

let rec map f list =
     match list with
     | [] -> []
     | h :: t -> f h :: map f t

let rec filter f list =
     match list with
     | [] -> []
     | h :: t -> if f h then  h :: filter f t else filter f t
 
let append xs ys =
    let rec revappend _xs _ys =
        match _xs with
        | [] -> _ys
        | h :: t -> revappend t (h :: _ys)
    let revxs = reverse xs
    revappend revxs ys

let rec concat xs =
    match xs with
    | [] -> []
    | h :: t -> append h (concat t)